src/analysis_and_optimization/Mem_pattern.ml

```diff
@@ -0,0 +1,602 @@
+ *  cell access. Returns true if it finds
+ * a vector, row vector, matrix, or matrix with single cell access
+ * as well as an array of any of the above that is accessing the 
+ * inner matrix types cell.
+ *)
+let rec is_uni_eigen_loop_indexing (ut : UnsizedType.t)
+    (index : 'a Expr.Fixed.t Index.t list) =
+  match (ut, index) with
+  | (UnsizedType.UArray t | UFun (_, ReturnType t, _, _)), index -> (
+    match List.tl index with
+    | Some cut_list -> is_uni_eigen_loop_indexing t cut_list
+    | None -> is_uni_eigen_loop_indexing t [Index.All] )
+  | (UVector | URowVector), [Index.Single (_ : 'a Expr.Fixed.t)] -> true
+  | UMatrix, [Single (_ : 'a Expr.Fixed.t); Single (_ : 'a Expr.Fixed.t)] ->
+      true
+  (* None of the below contain single cell access*)
+  | (UReal | UInt | UMathLibraryFunction | UFun (_, Void, _, _)), _ -> false
+  | (UVector | URowVector), _ -> false
+  | UMatrix, _ -> false
+
+(**
+ * Specialization to find demotable (SoA -> AoS) objects
+ *  that carries logic on whether we are in a for loop or not.
+ * (TODO: Could be combined with other query demotable above.)
+ *)
+let rec query_demotable_exprs_loop in_loop Expr.Fixed.({pattern; _}) =
+  let query_expr = query_demotable_exprs_loop in_loop in
+  match pattern with
+  | FunApp (kind, (exprs : Expr.Typed.Meta.t Expr.Fixed.t list)) ->
+      query_demotable_funkinds_loop query_demotable_exprs_loop in_loop kind
+        exprs
+  | Var (_ : string) | Lit ((_ : Expr.Fixed.Pattern.litType), (_ : string)) ->
+      Set.Poly.empty
+  | TernaryIf (predicate, texpr, fexpr) ->
+      Set.Poly.union
+        (Set.Poly.union (query_expr predicate) (query_expr texpr))
+        (query_expr fexpr)
+  | Indexed (Expr.Fixed.({pattern= Var name; meta= {type_; _}}), indexed) ->
+      let query_index = map_index query_expr in
+      if in_loop && is_uni_eigen_loop_indexing type_ indexed then
+        Set.Poly.union (Set.Poly.singleton name)
+          (Set.Poly.union_list (List.map ~f:query_index indexed))
+      else Set.Poly.union_list (List.map ~f:query_index indexed)
+  | Indexed (expr, indexed) ->
+      let query_index = map_index query_expr in
+      Set.Poly.union (query_expr expr)
+        (Set.Poly.union_list (List.map ~f:query_index indexed))
+  | EAnd (lhs, rhs) | EOr (lhs, rhs) ->
+      Set.Poly.union (query_expr lhs) (query_expr rhs)
+
+(**
+ * Specialization of the above to handle demotion while carrying 
+ * information on whether we have entered a for loop.
+ * (TODO: We can probably coerce this and the 
+ *  other query demotable function together)
+ * 
+ *)
+let rec query_demotable_stmts_loop in_loop Stmt.Fixed.({pattern; _}) =
+  let query_expr = query_demotable_exprs_loop in_loop in
+  let query_inner_stmt checking_loop =
+    query_demotable_stmts_loop checking_loop
+  in
+  match pattern with
+  | Stmt.Fixed.Pattern.Decl _ -> Set.Poly.empty
+  | Assignment (((name : string), (_ : UnsizedType.t), lhs), rhs) ->
```

Comment:
Do we need to check the assignment LHS for single indexing?

Reply:
Yes so in the 

```
      let check_lhs =
        Set.Poly.union_list (List.map ~f:(map_index query_expr) lhs)
      in
```

`query_expr`  is `let query_expr = query_demotable_exprs_loop in_loop` so if we currently in a loop we check that side as well

Reply:
So `lhs` here is the list of LHS indices, right? so doesn't this check if there are any bad single-indexed expressions in each of the LHS indices, but doesn't check if the LHS itself is a single index?